File: app/src/main/java/fr/gstraymond/ui/adapter/SearchViewCursorAdapter.java
package fr.gstraymond.ui.adapter;

import android.content.Context;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.graphics.drawable.Drawable;
import android.support.v4.content.res.ResourcesCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CursorAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.magic.card.search.commons.log.Log;

import java.util.ArrayList;
import java.util.List;

import fr.gstraymond.R;
import fr.gstraymond.autocomplete.response.Option;
import fr.gstraymond.autocomplete.response.Payload;
import fr.gstraymond.biz.SetImageGetter;
import fr.gstraymond.tools.CardColorUtil;


public class SearchViewCursorAdapter extends CursorAdapter {

    private static String FIELD_TYPE = "type";
    private static String FIELD_TEXT = "text";
    private static String FIELD_EDITION_CODE = "editionCode";
    private static String FIELD_CARD_COLOR_ID = "cardColorId";
    private static String FIELD_CARD_TYPE = "cardType";

    private static int FIELD_TYPE_TOKEN = 0;
    private static int FIELD_TYPE_EDITION = 1;
    private static int FIELD_TYPE_CARD = 2;

    private SetImageGetter setImageGetter;
    private Log log = new Log(this);

    public static SearchViewCursorAdapter empty(Context context) {
        return new SearchViewCursorAdapter(context, convert(new ArrayList<Option>()), 0);
    }

    public static SearchViewCursorAdapter from(Context context, List<Option> data) {
        return new SearchViewCursorAdapter(context, convert(data), 0);
    }

    private SearchViewCursorAdapter(Context context, Cursor cursor, int flags) {
        super(context, cursor, flags);
        setImageGetter = new SetImageGetter(context);
    }

    @Override
    public View newView(Context context, Cursor cursor, ViewGroup parent) {
        ViewHolder holder;
        View v;
        LayoutInflater inflater = LayoutInflater.from(context);

        int type = getType(cursor);
        if (type == FIELD_TYPE_EDITION) {
            v = inflater.inflate(R.layout.searchview_adapter_edition, parent, false);
            holder = new EditionViewHolder(
                    (TextView) v.findViewById(R.id.searchview_edition_text),
                    (ImageView) v.findViewById(R.id.searchview_edition_image)
            );
        } else if (type == FIELD_TYPE_CARD) {
            v = inflater.inflate(R.layout.searchview_adapter_card, parent, false);
            holder = new CardViewHolder(
                    (TextView) v.findViewById(R.id.searchview_card_title),
                    (TextView) v.findViewById(R.id.searchview_card_type)
            );
        } else {
            v = inflater.inflate(R.layout.searchview_adapter_token, parent, false);
            holder = new TokenViewHolder((TextView) v);
        }

        v.setTag(holder);
        return v;
    }

    @Override
    public void bindView(View view, Context context, Cursor cursor) {
        ViewHolder holder = (ViewHolder) view.getTag();
        String text = cursor.getString(cursor.getColumnIndex(FIELD_TEXT));

        if (holder instanceof EditionViewHolder) {
            EditionViewHolder editionViewHolder = (EditionViewHolder) holder;
            editionViewHolder.textView.setText(text);
            Drawable drawable = setImageGetter.getDrawable(getEditionCode(cursor));
            editionViewHolder.imageView.setImageDrawable(drawable);
        }

        if (holder instanceof CardViewHolder) {
            CardViewHolder cardViewHolder = (CardViewHolder) holder;
            int colorId = ResourcesCompat.getColor(context.getResources(), getCardColorId(cursor), null);
            cardViewHolder.textViewTitle.setTextColor(colorId);
            cardViewHolder.textViewTitle.setText(text);
            cardViewHolder.textViewType.setText(getCardType(cursor));
        }

        if (holder instanceof TokenViewHolder) {
            TokenViewHolder tokenViewHolder = (TokenViewHolder) holder;
            tokenViewHolder.textView.setText(text);
        }
    }

    interface ViewHolder {
    }

    private static class TokenViewHolder implements ViewHolder {
        TextView textView;

        TokenViewHolder(TextView textView) {
            this.textView = textView;
        }
    }

    private static class EditionViewHolder implements ViewHolder {
        TextView textView;
        ImageView imageView;

        EditionViewHolder(TextView textView, ImageView imageView) {
            this.textView = textView;
            this.imageView = imageView;
        }
    }

    private static class CardViewHolder implements ViewHolder {
        TextView textViewTitle;
        TextView textViewType;

        CardViewHolder(TextView textViewTitle, TextView textViewType) {
            this.textViewTitle = textViewTitle;
            this.textViewType = textViewType;
        }
    }

    @Override
    public int getItemViewType(int position) {
        Cursor cursor = (Cursor) getItem(position);
        return getType(cursor);
    }

    @Override
    public int getViewTypeCount() {
        return 3;
    }

    public void changeCursor(List<Option> data) {
        super.changeCursor(convert(data));
    }

    private static Cursor convert(List<Option> data) {
        MatrixCursor cursor = new MatrixCursor(
                new String[]{
                        "_id",
                        FIELD_TYPE,
                        FIELD_TEXT,
                        FIELD_EDITION_CODE,
                        FIELD_CARD_COLOR_ID,
                        FIELD_CARD_TYPE
                }
        );

        for (int i = 0; i < data.size(); i++) {
            Option option = data.get(i);
            int type = FIELD_TYPE_TOKEN;
            String editionCode = null;
            Integer cardColorId = null;
            String cardType = null;
            Payload payload = option.getPayload();
            if (payload != null) {
                if (payload.getStdEditionCode() != null) {
                    type = FIELD_TYPE_EDITION;
                    editionCode = payload.getStdEditionCode();
                } else {
                    type = FIELD_TYPE_CARD;
                    cardColorId =  CardColorUtil.getColorId(payload.getColors(), payload.getType());
                    cardType = payload.getType();
                }
            }
            cursor.addRow(new Object[]{
                    i,
                    type,
                    option.getText(),
                    editionCode,
                    cardColorId,
                    cardType
            });
        }
        return cursor;
    }

    private String getEditionCode(Cursor cursor) {
        return cursor.getString(cursor.getColumnIndex(FIELD_EDITION_CODE));
    }

    private int getType(Cursor cursor) {
        return cursor.getInt(cursor.getColumnIndex(FIELD_TYPE));
    }

    private int getCardColorId(Cursor cursor) {
        return cursor.getInt(cursor.getColumnIndex(FIELD_CARD_COLOR_ID));
    }

    private String getCardType(Cursor cursor) {
        return cursor.getString(cursor.getColumnIndex(FIELD_CARD_TYPE));
    }
}

File: app/src/main/kotlin/fr/gstraymond/ui/adapter/SearchViewCursorAdapter.kt
package fr.gstraymond.ui.adapter

import android.content.Context
import android.database.Cursor
import android.database.MatrixCursor
import android.support.v4.content.res.ResourcesCompat
import android.support.v4.widget.CursorAdapter
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import fr.gstraymond.R
import fr.gstraymond.biz.SetImageGetter
import fr.gstraymond.models.autocomplete.response.Option
import fr.gstraymond.tools.CardColorUtil
import fr.gstraymond.utils.inflate
import java.util.*

class SearchViewCursorAdapter private constructor(context: Context,
                                                  cursor: Cursor,
                                                  flags: Int) : CursorAdapter(context, cursor, flags) {

    private val setImageGetter = SetImageGetter(context)

    override fun newView(context: Context, cursor: Cursor, parent: ViewGroup): View {
        val holder: ViewHolder
        val view: View

        when (getType(cursor)) {
            FIELD_TYPE_EDITION -> {
                view = context.inflate(R.layout.searchview_adapter_edition, parent)
                holder = EditionViewHolder(
                        view.findViewById(R.id.searchview_edition_text) as TextView,
                        view.findViewById(R.id.searchview_edition_image) as ImageView)
            }
            FIELD_TYPE_CARD -> {
                view = context.inflate(R.layout.searchview_adapter_card, parent)
                holder = CardViewHolder(
                        view.findViewById(R.id.searchview_card_title) as TextView,
                        view.findViewById(R.id.searchview_card_type) as TextView)
            }
            else -> {
                view = context.inflate(R.layout.searchview_adapter_token, parent)
                holder = TokenViewHolder(view as TextView)
            }
        }

        return view.apply { tag = holder }
    }

    override fun bindView(view: View, context: Context, cursor: Cursor) {
        val holder = view.tag as ViewHolder
        val text = cursor.getString(cursor.getColumnIndex(FIELD_TEXT))

        when (holder) {
            is EditionViewHolder -> {
                holder.textView.text = text
                val drawable = setImageGetter.getDrawable(getEditionCode(cursor))
                holder.imageView.setImageDrawable(drawable)
            }

            is CardViewHolder -> {
                val colorId = ResourcesCompat.getColor(context.resources, getCardColorId(cursor), null)
                holder.textViewTitle.setTextColor(colorId)
                holder.textViewTitle.text = text
                holder.textViewType.text = getCardType(cursor)
            }

            is TokenViewHolder -> {
                holder.textView.text = text
            }
        }
    }

    interface ViewHolder

    class TokenViewHolder(var textView: TextView) : ViewHolder

    class EditionViewHolder(var textView: TextView,
                            var imageView: ImageView) : ViewHolder

    class CardViewHolder(var textViewTitle: TextView,
                         var textViewType: TextView) : ViewHolder

    override fun getItemViewType(position: Int): Int {
        val cursor = getItem(position) as Cursor
        return getType(cursor)
    }

    override fun getViewTypeCount() = 3

    fun changeCursor(data: List<Option>) {
        super.changeCursor(convert(data))
    }

    private fun getEditionCode(cursor: Cursor) =
            cursor.getString(cursor.getColumnIndex(FIELD_EDITION_CODE))

    private fun getType(cursor: Cursor) =
            cursor.getInt(cursor.getColumnIndex(FIELD_TYPE))

    private fun getCardColorId(cursor: Cursor) =
            cursor.getInt(cursor.getColumnIndex(FIELD_CARD_COLOR_ID))

    private fun getCardType(cursor: Cursor) =
            cursor.getString(cursor.getColumnIndex(FIELD_CARD_TYPE))

    companion object {

        private val FIELD_TYPE = "type"
        private val FIELD_TEXT = "text"
        private val FIELD_EDITION_CODE = "editionCode"
        private val FIELD_CARD_COLOR_ID = "cardColorId"
        private val FIELD_CARD_TYPE = "cardType"

        private val FIELD_TYPE_TOKEN = 0
        private val FIELD_TYPE_EDITION = 1
        private val FIELD_TYPE_CARD = 2

        fun empty(context: Context): SearchViewCursorAdapter {
            return SearchViewCursorAdapter(context, convert(ArrayList<Option>()), 0)
        }

        private fun convert(data: List<Option>): Cursor {
            val cursor = MatrixCursor(
                    arrayOf("_id",
                            FIELD_TYPE,
                            FIELD_TEXT,
                            FIELD_EDITION_CODE,
                            FIELD_CARD_COLOR_ID,
                            FIELD_CARD_TYPE
                    )
            )

            data.forEachIndexed { i, option ->
                var type = FIELD_TYPE_TOKEN
                var editionCode: String? = null
                var cardColorId: Int? = null
                var cardType: String? = null

                option.payload?.apply {
                    if (stdEditionCode != null) {
                        type = FIELD_TYPE_EDITION
                        editionCode = stdEditionCode
                    } else {
                        type = FIELD_TYPE_CARD
                        cardColorId = CardColorUtil.getColorId(colors, this.type)
                        cardType = this.type
                    }
                }
                cursor.addRow(
                        arrayOf(i,
                                type,
                                option.text,
                                editionCode,
                                cardColorId,
                                cardType
                        )
                )
            }
            return cursor
        }
    }
}
